File: UnmanagedToManaged/Engine/Transform/NativeSymbolTransform.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Imports System.Text.RegularExpressions

Namespace Transform

    ''' <summary>
    ''' Runs certain transformations on symbols
    ''' </summary>
    ''' <remarks></remarks>
    Public Class NativeSymbolTransform

        Private m_it As New NativeSymbolIterator()

        Public Sub New()
        End Sub

        Public Sub CollapseNamedTypes(ByVal ns As NativeSymbol)
            For Each rel As NativeSymbolRelationship In m_it.FindAllNativeSymbolRelationships(ns)
                CollapseNamedTypesImpl(rel.Parent, rel.Symbol)
            Next
        End Sub

        Private Sub CollapseNamedTypesImpl(ByVal ns As NativeSymbol, ByVal child As NativeSymbol)
            If ns Is Nothing Then
                Return
            End If
            ThrowIfNull(child)

            If child.Kind = NativeSymbolKind.NamedType Then
                Dim namedNt As NativeNamedType = DirectCast(child, NativeNamedType)
                If namedNt.RealType IsNot Nothing Then
                    ns.ReplaceChild(child, namedNt.RealType)
                End If
            End If
        End Sub

        Public Sub CollapseTypedefs(ByVal ns As NativeSymbol)
            For Each rel As NativeSymbolRelationship In m_it.FindAllNativeSymbolRelationships(ns)
                CollapseTypedefsImpl(rel.Parent, rel.Symbol)
            Next
        End Sub

        Private Sub CollapseTypedefsImpl(ByVal ns As NativeSymbol, ByVal child As NativeSymbol)
            If ns Is Nothing Then
                Return
            End If
            ThrowIfNull(child)

            If child.Kind = NativeSymbolKind.TypedefType Then
                Dim typedef As NativeTypeDef = DirectCast(child, NativeTypeDef)
                If typedef.RealType IsNot Nothing Then
                    ns.ReplaceChild(child, typedef.RealType)
                End If
            End If
        End Sub

        ''' <summary>
        ''' Renames matching defined types and named types to the new name
        ''' </summary>
        ''' <param name="ns"></param>
        ''' <param name="oldName"></param>
        ''' <param name="newName"></param>
        ''' <remarks></remarks>
        Public Sub RenameTypeSymbol(ByVal ns As NativeSymbol, ByVal oldName As String, ByVal newName As String)
            For Each sym As NativeSymbol In m_it.FindAllNativeSymbols(ns)
                If (sym.Category = NativeSymbolCategory.Defined OrElse sym.Kind = NativeSymbolKind.NamedType) _
                    AndAlso 0 = String.CompareOrdinal(sym.Name, oldName) Then
                    sym.Name = newName
                End If
            Next
        End Sub

        ''' <summary>
        ''' Inspect the type name and determine if there is a better name for it 
        ''' </summary>
        ''' <param name="definedNt"></param>
        ''' <remarks></remarks>
        Public Sub RunTypeNameHeuristics(ByVal definedNt As NativeDefinedType)
            ThrowIfNull(definedNt)


        End Sub

    End Class

End Namespace

File: PInvoke.Core/Transform/NativeSymbolTransform.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using PInvoke.NativeTypes;
using PInvoke.NativeTypes.Enums;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Text.RegularExpressions;
using static PInvoke.Contract;

namespace PInvoke.Transform
{
    /// <summary>
    /// Runs certain transformations on symbols
    /// </summary>
    /// <remarks></remarks>
    public class NativeSymbolTransform
    {
        private NativeSymbolIterator iterator = new NativeSymbolIterator();

        public NativeSymbolTransform()
        {
        }

        public void CollapseNamedTypes(NativeSymbol ns)
        {
            foreach (var relationship in iterator.FindAllNativeSymbolRelationships(ns))
            {
                CollapseNamedTypesImpl(relationship.Parent, relationship.Symbol);
            }
        }

        private void CollapseNamedTypesImpl(NativeSymbol ns, NativeSymbol child)
        {
            if (ns == null)
            {
                return;
            }
            ThrowIfNull(child);

            if (child.Kind == NativeSymbolKind.NamedType)
            {
                var namedNt = (NativeNamedType)child;
                if (namedNt.RealType != null)
                {
                    ns.ReplaceChild(child, namedNt.RealType);
                }
            }
        }

        public void CollapseTypedefs(NativeSymbol ns)
        {
            foreach (var relationship in iterator.FindAllNativeSymbolRelationships(ns))
            {
                CollapseTypedefsImpl(relationship.Parent, relationship.Symbol);
            }
        }

        private void CollapseTypedefsImpl(NativeSymbol ns, NativeSymbol child)
        {
            if (ns == null)
            {
                return;
            }
            ThrowIfNull(child);

            if (child.Kind == NativeSymbolKind.TypeDefType)
            {
                var typedef = (NativeTypeDef)child;
                if (typedef.RealType != null)
                {
                    ns.ReplaceChild(child, typedef.RealType);
                }
            }
        }

        /// <summary>
        /// Renames matching defined types and named types to the new name
        /// </summary>
        /// <param name="ns"></param>
        /// <param name="oldName"></param>
        /// <param name="newName"></param>
        /// <remarks></remarks>
        public void RenameTypeSymbol(NativeSymbol ns, string oldName, string newName)
        {
            foreach (var sym in iterator.FindAllNativeSymbols(ns))
            {
                if ((sym.Category == NativeSymbolCategory.Defined || sym.Kind == NativeSymbolKind.NamedType) && 0 == string.CompareOrdinal(sym.Name, oldName))
                {
                    sym.Name = newName;
                }
            }
        }

        /// <summary>
        /// Inspect the type name and determine if there is a better name for it 
        /// </summary>
        /// <param name="definedNt"></param>
        /// <remarks></remarks>
        public void RunTypeNameHeuristics(NativeDefinedType definedNt)
        {
            ThrowIfNull(definedNt);
        }
    }
}
